File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerSessionMiddleware.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.arrow.flight;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

/** Middleware for handling Flight SQL Sessions including session cookie handling. */
public class ServerSessionMiddleware implements FlightServerMiddleware {
  Factory factory;
  boolean existingSession;
  private Session session;
  private String closedSessionId = null;

  public static final String sessionCookieName = "arrow_flight_session_id";

  /** Factory for managing and accessing ServerSessionMiddleware. */
  public static class Factory implements FlightServerMiddleware.Factory<ServerSessionMiddleware> {
    private final ConcurrentMap<String, Session> sessionStore = new ConcurrentHashMap<>();
    private final Callable<String> idGenerator;

    /**
     * Construct a factory for ServerSessionMiddleware.
     *
     * <p>Factory manages and accesses persistent sessions based on HTTP cookies.
     *
     * @param idGenerator A Callable returning unique session id Strings.
     */
    public Factory(Callable<String> idGenerator) {
      this.idGenerator = idGenerator;
    }

    private synchronized Session createNewSession() {
      String id;
      try {
        id = idGenerator.call();
      } catch (Exception ignored) {
        // Most impls aren't going to throw so don't make caller handle a nonexistent checked
        // exception
        throw CallStatus.INTERNAL.withDescription("Session creation error").toRuntimeException();
      }

      Session newSession = new Session(id);
      if (sessionStore.putIfAbsent(id, newSession) != null) {
        // Collision, should never happen
        throw CallStatus.INTERNAL.withDescription("Session creation error").toRuntimeException();
      }
      return newSession;
    }

    private void closeSession(String id) {
      if (sessionStore.remove(id) == null) {
        throw CallStatus.NOT_FOUND
            .withDescription("Session id '" + id + "' not found.")
            .toRuntimeException();
      }
    }

    @Override
    public ServerSessionMiddleware onCallStarted(
        CallInfo callInfo, CallHeaders incomingHeaders, RequestContext context) {
      String sessionId = null;

      final Iterable<String> it = incomingHeaders.getAll("cookie");
      if (it != null) {
        findIdCookie:
        for (final String headerValue : it) {
          for (final String cookie : headerValue.split(" ;")) {
            final String[] cookiePair = cookie.split("=");
            if (cookiePair.length != 2) {
              // Soft failure:  Ignore invalid cookie list field
              break;
            }

            if (sessionCookieName.equals(cookiePair[0]) && cookiePair[1].length() > 0) {
              sessionId = cookiePair[1];
              break findIdCookie;
            }
          }
        }
      }

      if (sessionId == null) {
        // No session cookie, create middleware instance without session.
        return new ServerSessionMiddleware(this, incomingHeaders, null);
      }

      Session session = sessionStore.get(sessionId);
      // Cookie provided by caller, but invalid
      if (session == null) {
        // Can't soft-fail/proceed here, clients will get unexpected behaviour without options they
        // thought were set.
        throw CallStatus.NOT_FOUND
            .withDescription("Invalid " + sessionCookieName + " cookie.")
            .toRuntimeException();
      }

      return new ServerSessionMiddleware(this, incomingHeaders, session);
    }
  }

  /** A thread-safe container for named SessionOptionValues. */
  public static class Session {
    public final String id;
    private ConcurrentMap<String, SessionOptionValue> sessionData =
        new ConcurrentHashMap<String, SessionOptionValue>();

    /**
     * Construct a new Session with the given id.
     *
     * @param id The Session's id string, which is used as the session cookie value.
     */
    private Session(String id) {
      this.id = id;
    }

    /** Get session option by name, or null if it does not exist. */
    public SessionOptionValue getSessionOption(String name) {
      return sessionData.get(name);
    }

    /** Get an immutable copy of the session options map. */
    public Map<String, SessionOptionValue> getSessionOptions() {
      return Collections.unmodifiableMap(new HashMap(sessionData));
    }

    /** Set session option by name to given value. */
    public void setSessionOption(String name, SessionOptionValue value) {
      sessionData.put(name, value);
    }

    /** Idempotently remove name from this session. */
    public void eraseSessionOption(String name) {
      sessionData.remove(name);
    }
  }

  private final CallHeaders headers;

  private ServerSessionMiddleware(
      ServerSessionMiddleware.Factory factory, CallHeaders incomingHeaders, Session session) {
    this.factory = factory;
    headers = incomingHeaders;
    this.session = session;
    existingSession = (session != null);
  }

  /**
   * Check if there is an open session associated with this call.
   *
   * @return True iff there is an open session associated with this call.
   */
  public boolean hasSession() {
    return session != null;
  }

  /**
   * Get the existing or new session value map for this call.
   *
   * @return The session option value map, or null in case of an id generation collision.
   */
  public synchronized Session getSession() {
    if (session == null) {
      session = factory.createNewSession();
    }

    return session;
  }

  /**
   * Close the current session.
   *
   * <p>It is an error to call this without a valid session specified via cookie or equivalent.
   */
  public synchronized void closeSession() {
    if (session == null) {
      throw CallStatus.NOT_FOUND
          .withDescription("No session found for the current call.")
          .toRuntimeException();
    }
    factory.closeSession(session.id);
    closedSessionId = session.id;
    session = null;
  }

  public CallHeaders getCallHeaders() {
    return headers;
  }

  @Override
  public void onBeforeSendingHeaders(CallHeaders outgoingHeaders) {
    if (!existingSession && session != null) {
      outgoingHeaders.insert("set-cookie", sessionCookieName + "=" + session.id);
    }
    if (closedSessionId != null) {
      outgoingHeaders.insert(
          "set-cookie", sessionCookieName + "=" + closedSessionId + "; Max-Age=0");
    }
  }

  @Override
  public void onCallCompleted(CallStatus status) {}

  @Override
  public void onCallErrored(Throwable err) {}
}

File: cpp/src/arrow/flight/sql/server_session_middleware.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include <mutex>

#include "arrow/flight/sql/server_session_middleware.h"
#include "arrow/flight/sql/server_session_middleware_factory.h"

namespace arrow {
namespace flight {
namespace sql {

class ServerSessionMiddlewareImpl : public ServerSessionMiddleware {
 protected:
  std::shared_mutex mutex_;
  ServerSessionMiddlewareFactory* factory_;
  const CallHeaders& headers_;
  std::shared_ptr<FlightSession> session_;
  std::string session_id_;
  std::string closed_session_id_;
  bool existing_session_;

 public:
  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
                              const CallHeaders& headers)
      : factory_(factory), headers_(headers), existing_session_(false) {}

  ServerSessionMiddlewareImpl(ServerSessionMiddlewareFactory* factory,
                              const CallHeaders& headers,
                              std::shared_ptr<FlightSession> session,
                              std::string session_id, bool existing_session = true)
      : factory_(factory),
        headers_(headers),
        session_(std::move(session)),
        session_id_(std::move(session_id)),
        existing_session_(existing_session) {}

  void SendingHeaders(AddCallHeaders* add_call_headers) override {
    if (!existing_session_ && session_) {
      add_call_headers->AddHeader(
          "set-cookie", static_cast<std::string>(kSessionCookieName) + "=" + session_id_);
    }
    if (!closed_session_id_.empty()) {
      add_call_headers->AddHeader(
          "set-cookie", static_cast<std::string>(kSessionCookieName) + "=" + session_id_ +
                            "; Max-Age=0");
    }
  }

  void CallCompleted(const Status&) override {}

  bool HasSession() const override { return static_cast<bool>(session_); }

  arrow::Result<std::shared_ptr<FlightSession>> GetSession() override {
    const std::lock_guard<std::shared_mutex> l(mutex_);
    if (!session_) {
      auto [id, s] = factory_->CreateNewSession();
      session_ = std::move(s);
      session_id_ = std::move(id);
    }
    if (!static_cast<bool>(session_)) {
      return Status::UnknownError("Error creating session.");
    }
    return session_;
  }

  Status CloseSession() override {
    const std::lock_guard<std::shared_mutex> l(mutex_);
    if (!static_cast<bool>(session_)) {
      return Status::Invalid("Nonexistent session cannot be closed.");
    }
    ARROW_RETURN_NOT_OK(factory_->CloseSession(session_id_));
    closed_session_id_ = std::move(session_id_);
    session_id_.clear();
    session_.reset();
    existing_session_ = false;

    return Status::OK();
  }

  const CallHeaders& GetCallHeaders() const override { return headers_; }
};

std::vector<std::pair<std::string, std::string>>
ServerSessionMiddlewareFactory::ParseCookieString(const std::string_view& s) {
  const std::string list_sep = "; ";
  const std::string pair_sep = "=";

  std::vector<std::pair<std::string, std::string>> result;

  size_t cur = 0;
  while (cur < s.length()) {
    const size_t end = s.find(list_sep, cur);
    const bool further_pairs = end != std::string::npos;
    const size_t len = further_pairs ? end - cur : std::string::npos;
    const std::string_view tok = s.substr(cur, len);
    cur = further_pairs ? end + list_sep.length() : s.length();

    const size_t val_pos = tok.find(pair_sep);
    if (val_pos == std::string::npos) {
      // The cookie header is somewhat malformed; ignore the key and continue parsing
      continue;
    }
    const std::string_view cookie_name = tok.substr(0, val_pos);
    std::string_view cookie_value =
        tok.substr(val_pos + pair_sep.length(), std::string::npos);
    if (cookie_name.empty()) {
      continue;
    }
    // Strip doublequotes
    if (cookie_value.length() >= 2 && cookie_value.front() == '"' &&
        cookie_value.back() == '"') {
      cookie_value.remove_prefix(1);
      cookie_value.remove_suffix(1);
    }
    result.emplace_back(cookie_name, cookie_value);
  }

  return result;
}

Status ServerSessionMiddlewareFactory::StartCall(
    const CallInfo&, const CallHeaders& incoming_headers,
    std::shared_ptr<ServerMiddleware>* middleware) {
  std::string session_id;

  const std::pair<CallHeaders::const_iterator, CallHeaders::const_iterator>&
      headers_it_pr = incoming_headers.equal_range("cookie");
  for (auto itr = headers_it_pr.first; itr != headers_it_pr.second; ++itr) {
    const std::string_view& cookie_header = itr->second;
    const std::vector<std::pair<std::string, std::string>> cookies =
        ParseCookieString(cookie_header);
    for (const std::pair<std::string, std::string>& cookie : cookies) {
      if (cookie.first == kSessionCookieName) {
        if (cookie.second.empty())
          return Status::Invalid("Empty ", kSessionCookieName, " cookie value.");
        session_id = std::move(cookie.second);
      }
    }
    if (!session_id.empty()) break;
  }

  if (session_id.empty()) {
    // No cookie was found
    // Temporary workaround until middleware handling fixed
    auto [id, s] = CreateNewSession();
    *middleware = std::make_shared<ServerSessionMiddlewareImpl>(this, incoming_headers,
                                                                std::move(s), id, false);
  } else {
    const std::shared_lock<std::shared_mutex> l(session_store_lock_);
    if (auto it = session_store_.find(session_id); it == session_store_.end()) {
      return Status::Invalid("Invalid or expired ", kSessionCookieName, " cookie.");
    } else {
      auto session = it->second;
      *middleware = std::make_shared<ServerSessionMiddlewareImpl>(
          this, incoming_headers, std::move(session), session_id);
    }
  }

  return Status::OK();
}

/// \brief Get a new, empty session option map & its id key; {"",NULLPTR} on collision.
std::pair<std::string, std::shared_ptr<FlightSession>>
ServerSessionMiddlewareFactory::CreateNewSession() {
  auto new_id = id_generator_();
  auto session = std::make_shared<FlightSession>();

  const std::lock_guard<std::shared_mutex> l(session_store_lock_);
  if (session_store_.count(new_id)) {
    // Collision
    return {"", NULLPTR};
  }
  session_store_[new_id] = session;

  return {new_id, session};
}

Status ServerSessionMiddlewareFactory::CloseSession(std::string id) {
  const std::lock_guard<std::shared_mutex> l(session_store_lock_);
  if (!session_store_.erase(id)) {
    return Status::KeyError("Invalid or nonexistent session cannot be closed.");
  }
  return Status::OK();
}

std::shared_ptr<ServerMiddlewareFactory> MakeServerSessionMiddlewareFactory(
    std::function<std::string()> id_gen) {
  return std::make_shared<ServerSessionMiddlewareFactory>(std::move(id_gen));
}

std::optional<SessionOptionValue> FlightSession::GetSessionOption(
    const std::string& name) {
  const std::shared_lock<std::shared_mutex> l(map_lock_);
  auto it = map_.find(name);
  if (it != map_.end()) {
    return it->second;
  } else {
    return std::nullopt;
  }
}

std::map<std::string, SessionOptionValue> FlightSession::GetSessionOptions() {
  const std::shared_lock<std::shared_mutex> l(map_lock_);
  return map_;
}

void FlightSession::SetSessionOption(const std::string& name,
                                     const SessionOptionValue value) {
  const std::lock_guard<std::shared_mutex> l(map_lock_);
  map_[name] = std::move(value);
}

void FlightSession::EraseSessionOption(const std::string& name) {
  const std::lock_guard<std::shared_mutex> l(map_lock_);
  map_.erase(name);
}

}  // namespace sql
}  // namespace flight
}  // namespace arrow
